<a href="https://colab.research.google.com/github/GIvanConder/inacdoc/blob/main/Actividad3BDManipulacion_A01840602.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC5053 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 3**
Bases, almacenes y manipulación de datos

---

*   NOMBRE: Gerardo Iván Conde Reynoso
*   MATRÍCULA: A01840602


---

En esta actividad usarás una base de datos relacional basada en el informe de participación y la lista del top 10 de Netflix. Incluye películas y programas de televisión, así como información sobre temporadas, métricas de visualización, fechas de estreno, duración y más, organizada en las siguientes tablas:

* `movie`: Información general de las películas.

* `tv_show`: Información general de los programas de televisión.

* `season`: Datos de las temporadas asociadas a cada programa de TV.

* `view_summary`: Métricas de visualización y rendimiento de películas o temporadas.

Revisa con detalle su esquema para que comprendas cómo se relacionan las tablas anteriores.

**NOTA IMPORTANTE:** Asegúrate de responder *explícitamente* todos los cuestionamientos.


`PyMySQL` es una librería escrita en Python puro que funciona como conector (*driver*) para motores de bases de datos MySQL, permitiendo abrir conexiones, ejecutar consultas SQL y recuperar resultados directamente desde programas en Python.

In [21]:
pip install pymysql

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.3/45.3 kB 1.9 MB/s eta 0:00:00


`SQLAlchemy` es una librería de Python que facilita la interacción con bases de datos y permite mantener un pool de conexiones eficiente, gestionar *commits* y *rollbacks* de forma automática y asegurar que múltiples conexiones simultáneas se manejen de manera segura, incluso cuando se ejecutan consultas SQL “puras”

In [22]:
# Importa las librerías necesarias
import pymysql
import sqlalchemy as sqla
import pandas as pd
import datetime as dt

Se crea una conexión (`conn`) para luego invocar declaraciones SQL.

In [23]:
# motor+driver://usuarioBD:clave@ipHostDBMS:puerto/esquema
# pool_recycle controla el tiempo máximo de vida de una conexión en el pool (3600 segundos = 1 hora)
db = sqla.create_engine('mysql+pymysql://mnaTC4029User:mnaTC4029Pass!@172.208.104.202:3306/Netflix', pool_recycle=360000000000)
conn = db.connect()

Para que tus consultas sean más legibles y fáciles de mantener, puedes usar este formato multilínea con comillas triples y `sqla.text()`. Por ejemplo:

```
query = sqla.text("""
  SELECT ---
  FROM ---
  WHERE ---
""")
pd.read_sql(query, conn)
```

In [24]:
query = sqla.text("""
    SELECT *
    FROM movie
""")

pd.read_sql(query, conn)

id created_date modified_date available_globally locale  \
0          1   2024-01-01    2024-01-01            b'\x01'   None   
1          2   2024-01-01    2024-01-01            b'\x01'   None   
2          3   2024-01-01    2024-01-01            b'\x01'   None   
3          4   2024-01-01    2024-01-01            b'\x01'   None   
4          5   2024-01-01    2024-01-01            b'\x00'   None   
...      ...          ...           ...                ...    ...   
11782  11783   2024-01-01    2024-01-01               None     en   
11783  11784   2024-01-01    2024-01-01               None     en   
11784  11785   2024-01-01    2024-01-01               None     en   
11785  11786   2024-01-01    2024-01-01               None     en   
11786  11787   2024-01-01    2024-01-01               None   None   

                original_title release_date  runtime  \
0                         None   2024-03-08      110   
1                         None   2024-01-12      107   
2      La sociedad de la nieve   2024-01-04      146   
3                Sous la Seine   2024-06-05      104   
4                         None         None       92   
...                        ...          ...      ...   
11782                     None         None      121   
11783                     None         None       97   
11784                     None         None      106   
11785                     None         None      153   
11786                     None         None       94   

                             title  
0                           Damsel  
1                             Lift  
2              Society of the Snow  
3                      Under Paris  
4      The Super Mario Bros. Movie  
...                            ...  
11782                     47 Ronin  
11783        The Angry Birds Movie  
11784                    Abduction  
11785                    Prisoners  
11786                 Mother's Day  

[11787 rows x 9 columns]

1.	Extrae toda la información de las películas que duran más de 5 horas.

In [25]:
query = sqla.text("""
    SELECT *
    FROM movie
    WHERE runtime > 300
""")

pd.read_sql(query, conn)

id created_date modified_date available_globally locale  \
0  5793   2024-01-01    2024-01-01            b'\x00'   None   
1  5794   2024-01-01    2024-01-01            b'\x00'   None   
2  5874   2024-01-01    2024-01-01            b'\x01'   None   
3  9729   2024-01-01    2024-01-01            b'\x00'   None   
4  9730   2024-01-01    2024-01-01            b'\x00'   None   

           original_title release_date  runtime  \
0        日本統一シリーズ: 映画シリーズ         None     3892   
1          釣りバカ日誌: 映画シリーズ         None     2120   
2                    None   2021-08-06      312   
3      織田同志会 織田征仁: 映画シリーズ         None      710   
4  キングダム～首領になった男～: 映画シリーズ         None      427   

                                               title  
0                   Nihontouitsu Series: Film Series  
1                  Free and Easy Series: Film Series  
2                           Navarasa: Limited Series  
3                             Seiji Oda: Film Series  
4  Kingdom ~ The Man Who Became the Top ~: Film S...

2.	¿Cuál es el porcentaje de películas disponibles únicamente en EU en relación con el total, excluyendo los valores `NULL`? La consulta SQL debe entregar directamente el porcentaje final. No se deben devolver resultados parciales para realizar el cálculo en Python.

In [26]:
query = sqla.text("""
    SELECT
        (SUM(CASE
            WHEN available_globally = 0 THEN 1
            ELSE 0
        END) * 100.0 / COUNT(*)) AS porcentaje_solo_eu
    FROM movie
    WHERE available_globally IS NOT NULL
""")

pd.read_sql(query, conn)

porcentaje_solo_eu
0            83.57175

3.	¿Cuáles son los idiomas o regiones originales en la tabla de películas?
* ¿Cuántos registros tienen el campo `locale` con valor `NULL`? (NULL en SQL ⇔ None en Python)

In [27]:
query = sqla.text("""
    SELECT DISTINCT locale
    FROM movie
""")

pd.read_sql(query, conn)

locale
0   None
1     en

In [28]:
query = sqla.text("""
    SELECT COUNT(*) AS total_locale_null
    FROM movie
    WHERE locale IS NULL
""")

pd.read_sql(query, conn)

total_locale_null
0              11255

4.	Asumiendo que los valores `NULL` en `locale` corresponden a otro idioma (diferente del inglés), el título original de la película NO debería coincidir con el título principal en dichos registros.
* Determina cuántas películas tienen títulos diferentes en estos dos campos (`title` y `original_title`).
*  ¿Coinciden los resultados (cantidad de `NULL` y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.
* Finalmente, concluye si la suposición de que los valores `NULL` en `locale` indican que la película está en otro idioma es válida.

In [29]:
query = sqla.text("""
    SELECT COUNT(*) AS total_titulos_diferentes
    FROM movie
    WHERE locale IS NULL
      AND title <> original_title
""")

pd.read_sql(query, conn)

total_titulos_diferentes
0                      3947

In [30]:
query = sqla.text("""
    SELECT
        title,
        original_title,
        locale
    FROM movie
    WHERE locale IS NULL
      AND title <> original_title
""")

pd.read_sql(query, conn)

title           original_title locale
0                   Society of the Snow  La sociedad de la nieve   None
1                           Under Paris            Sous la Seine   None
2                       Badland Hunters                       황야   None
3                         The Tearsmith   Fabbricante di lacrime   None
4                             The Abyss                Avgrunden   None
...                                 ...                      ...    ...
3942                           Wretches                      괴물들   None
3943  Yellow Door: '90s Lo-fi Film Club        노란문: 세기말 시네필 다이어리   None
3944        You Are the Apple of My Eye             那些年，我們一起追的女孩   None
3945            You're Everything To Me    Sen Benim HerŞeyimsin   None
3946                           Zone 261       Operation Ragnarok   None

[3947 rows x 3 columns]

In [31]:
query = sqla.text("""
    SELECT
        title,
        original_title,
        locale,
        available_globally
    FROM movie
    WHERE locale IS NULL
      AND title = original_title
""")

pd.read_sql(query, conn)

title                  original_title locale  \
0                   Recep Ivedik 7                  Recep İvedik 7   None   
1                         Veronica                        Verónica   None   
2                         Fantomas                        Fantômas   None   
3   Luccas Neto em: O Hotel Magico  Luccas Neto em: O Hotel Mágico   None   
4             Un Dia En El Paraiso            Un día en el paraíso   None   
5                       Superlopez                      Superlópez   None   
6             Baris Akarsu Merhaba            Barış Akarsu Merhaba   None   
7                            Totem                           Tótem   None   
8                   La Pena Maxima                  La pena máxima   None   
9                       Mirciulica                      Mirciulică   None   
10                          Animas                          Ánimas   None   
11                    Dejame Vivir                    Déjame Vivir   None   
12            Gangster Ka: African            Gangster Ka: Afričan   None   
13                      Party Hard                      Párty Hárd   None   
14                         Extasis                         Éxtasis   None   
15                       Malefique                       Maléfique   None   
16                        Tiramisu                        Tiramisù   None   

   available_globally  
0             b'\x00'  
1             b'\x00'  
2             b'\x00'  
3             b'\x00'  
4             b'\x00'  
5             b'\x01'  
6             b'\x00'  
7             b'\x00'  
8             b'\x01'  
9             b'\x00'  
10            b'\x00'  
11            b'\x00'  
12            b'\x00'  
13            b'\x00'  
14            b'\x00'  
15            b'\x00'  
16            b'\x00'

In [32]:
query = sqla.text("""
    SELECT COUNT(*) AS total_titulos_iguales
    FROM movie
    WHERE locale IS NULL
      AND title = original_title
""")

pd.read_sql(query, conn)

total_titulos_iguales
0                     17

In [33]:
query = sqla.text("""
    SELECT COUNT(*) AS titulos_con_null
    FROM movie
    WHERE locale IS NULL
      AND (title IS NULL OR original_title IS NULL)
""")
pd.read_sql(query, conn)


titulos_con_null
0              7291

Después de analizar los datos obtenidos en las consultas, se observa que la cantidad de resultados con valor NULL en locale es de 11255, de estos registros, la cantidad con titulos diferentes, es de 3947, para afinar las diferencias se calculó la cantidad de registros con titulos iguales los cuales representaron un total de 17 registros, por lo que se tiene una diferencia de 7291 títulos, para encontrar qué condición es la que realiza esta discrepancia del total de registros, se abordó la situación en que el valor de los registros de title o original_title podrían tener ausencia de valores (NULL), una vez que se aplicó esta consulta se obtuvo el valor de 7291, al encontrar este valor se pudo identificar el causante de que no coincidieran los resultados de las primeras consultas.

Con este análisis se llegó a la conlcusión de que las comparaciones con valores NULL no evalúan como verdaderas ni falsas, por lo que dichos registros quedan excluidos de ambas condiciones. Esto evidencia problemas de completitud en los datos y refuerza que la ausencia de locale no puede interpretarse directamente como un indicador de idioma distinto, si no como una característica que se debe de omitir, en adición a esto, en la visualización de los registros de películas con los títulos iguales, se observó que las diferencias son por reglas gramaticales propias de los idiomas originales, como tildes o caracteres especiales, esto tambíen añade que estas películas pueden estar en otros idiomas, por lo que esa comparación de títulos tampoco valida que sean exclusivas del idioma registrado y por lo tanto,  se debe de formular una nueva hipótesis para determinar si la película está en otro idioma.



5.	Determina el título de la película que ha permanecido más tiempo en el top 10.

In [35]:
query = sqla.text("""
    SELECT
        m.title,
        SUM(v.cumulative_weeks_in_top10) AS total_weeks_top10
    FROM view_summary v
    JOIN movie m
        ON v.movie_id = m.id
    GROUP BY m.title
    ORDER BY total_weeks_top10 DESC
    LIMIT 1
""")

pd.read_sql(query, conn)

title  total_weeks_top10
0  The Super Mario Bros. Movie              231.0

6.	Identifica los 5 programas de TV con mayor cantidad de temporadas.

In [36]:
query = sqla.text("""
    SELECT
        t.title,
        COUNT(*) AS total_seasons
    FROM season s
    JOIN tv_show t
        ON s.tv_show_id = t.id
    GROUP BY t.title
    ORDER BY total_seasons DESC
    LIMIT 5
""")

pd.read_sql(query, conn)

title  total_seasons
0                     Grey's Anatomy             20
1                   Naruto Shippuden             20
2                   Heartland (2007)             17
3  It's Always Sunny in Philadelphia             16
4                               NCIS             15

7.	¿Cuáles son los intervalos de fechas de los resúmenes en la tabla `view_summary` de los períodos (`duration`) semestrales?

In [37]:
query = sqla.text("""
    SELECT DISTINCT
        start_date,
        end_date
    FROM view_summary
    WHERE duration = 'semi_annually'
    ORDER BY start_date
""")

pd.read_sql(query, conn)

start_date    end_date
0  2023-07-01  2023-12-31
1  2024-01-01  2024-06-30

8.	Ordena las temporadas de *Grey's Anatomy* según la cantidad de vistas registradas en el primer período semestral de 2024.
* ¿Cómo interpretarías los resultados obtenidos?

In [38]:
query = sqla.text("""
    SELECT
        s.season_number,
        SUM(v.views) AS total_views
    FROM view_summary v
    JOIN season s
        ON v.season_id = s.id
    JOIN tv_show t
        ON s.tv_show_id = t.id
    WHERE t.title = 'Grey''s Anatomy'
      AND v.duration = 'semi_annually'
      AND v.start_date = '2024-01-01'
      AND v.end_date = '2024-06-30'
    GROUP BY s.season_number
    ORDER BY total_views DESC
""")

pd.read_sql(query, conn)

season_number  total_views
0               1    3600000.0
1               2    3100000.0
2               3    2900000.0
3               5    2900000.0
4               4    2900000.0
5               6    2800000.0
6               7    2700000.0
7               8    2600000.0
8               9    2500000.0
9              10    2400000.0
10             11    2200000.0
11             12    2000000.0
12             13    2000000.0
13             14    2000000.0
14             19    2000000.0
15             15    1900000.0
16             16    1800000.0
17             18    1700000.0
18             17    1700000.0
19             20     100000.0

El análisis del total de vistas por temporada de Grey’s Anatomy en el primer período semestral de 2024 muestra una tendencia clara en las primeras etapas de la serie. De la temporada 1 a la 14 se observa una disminución progresiva del interés, reflejada en una reducción gradual del número de vistas conforme avanzan las temporadas.

A partir de la temporada 15, el comportamiento de las vistas se vuelve irregular, sin una tendencia definida. Las temporadas posteriores presentan fluctuaciones que rompen el patrón observado previamente, lo que indica un consumo menos consistente en las etapas más recientes de la serie.

9.	Todas las consultas anteriores podrían hacerse también con la lógica relacional implementada en Pandas, que permite replicar la mayoría de las operaciones de SQL. Si los dataframes se han llenado como sigue, resuelve la consulta 8 utilizando las funciones de Pandas.

In [39]:
tv_show = pd.read_sql(sqla.text('SELECT * FROM tv_show'), conn)
season = pd.read_sql(sqla.text('SELECT * FROM season'), conn)
view_summary = pd.read_sql(sqla.text('SELECT * FROM view_summary'), conn)

In [41]:
greys = tv_show[tv_show['title'] == "Grey's Anatomy"]

greys_seasons = season.merge(
    greys,
    left_on='tv_show_id',
    right_on='id',
    how='inner'
)

vs_filtered = view_summary[
    (view_summary['duration'] == 'SEMI_ANNUALLY') &
    (view_summary['start_date'] >= dt.date(2024, 1, 1)) &
    (view_summary['end_date'] <= dt.date(2024, 6, 30))
]

greys_views = greys_seasons.merge(
    vs_filtered,
    left_on='id_x',
    right_on='season_id',
    how='inner'
)

resultado = (
    greys_views
    .groupby('season_number', as_index=False)['views']
    .sum()
    .rename(columns={'views': 'total_views'})
    .sort_values('total_views', ascending=False)
)

print(resultado)

    season_number  total_views
0             1.0      3600000
1             2.0      3100000
2             3.0      2900000
3             4.0      2900000
4             5.0      2900000
5             6.0      2800000
6             7.0      2700000
7             8.0      2600000
8             9.0      2500000
9            10.0      2400000
10           11.0      2200000
11           12.0      2000000
12           13.0      2000000
13           14.0      2000000
18           19.0      2000000
14           15.0      1900000
15           16.0      1800000
16           17.0      1700000
17           18.0      1700000
19           20.0       100000


`MySQL` es un sistema de gestión de bases de datos relacional, pero desde Python también es posible extraer información de bases de datos no relacionales, como `Firestore`, `MongoDB` o `Cassandra`, utilizando conectores o integraciones específicas. Esto permite combinar datos de diferentes fuentes según las necesidades del análisis o la aplicación.

En el siguiente ejercicio usarás un ejemplo con `Firestore` desde Python. Para ello utilizarás los módulos `credentials` y `firestore` de la biblioteca `firebase_admin`.

In [42]:
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore

En `Firestore`, a diferencia de `MySQL` donde se utiliza un usuario y contraseña para conectarse, la autenticación se realiza mediante un archivo JSON que almacena las credenciales necesarias para acceder a la base de datos. Este archivo contiene las claves y la información de configuración que permiten a Python establecer la conexión de manera segura.

In [43]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [44]:
# Debes descargar el archivo de credenciales "consultancy.json" (disponible en las instrucciones de Canvas) y ubicarte en la carpeta donde lo almacenaste
import os
DIR = "/content/drive/MyDrive/Colab Notebooks"
os.chdir(DIR)

In [45]:
# consultancy.json almacena la clave privada para autenticar una cuenta y autorizar el acceso a los servicios
# A través de la función Certificate(), se regresa una credencial inicializada, que puedes utilizar para crear una nueva instancia de la aplicación
cred = credentials.Certificate('consultancy.json')
firebase_admin.initialize_app(cred)
db = firestore.client()

10.	Investiga cómo leer la colección `EMPLOYEE` y mostrar su contenido en un dataframe. Asegúrate de incluir el `id` en el resultado

In [46]:
import pandas as pd
from IPython.display import display

employees_ref = db.collection('EMPLOYEE')

docs = employees_ref.stream()

employees = []

for doc in docs:
    data = doc.to_dict()
    data['id'] = doc.id
    employees.append(data)

df_employees = pd.DataFrame(employees)
df_employees = df_employees.set_index('id').reset_index()
display(df_employees)

id emp_fname   emp_lname              emp_hiredate
0  8LcLuxVHGAd2d9IQc5jF     David      Senior 1989-07-12 06:00:00+00:00
1  Fzd60D6Z2CU4n0wVV8YN   William  Smithfield 2004-06-04 05:00:00+00:00
2  lX5xuQ5w3i6ib2ExccWY      John        News 2000-11-08 06:00:00+00:00
3  yocFj2lichOkbAj9NBfp      June     Arbough 1996-12-01 06:00:00+00:00

In [20]:
firebase_admin.delete_app(firebase_admin.get_app())